treeherder/auth/backends.py

```diff
@@ -0,0 +1,148 @@
+import logging
+import re
+from hashlib import sha1
+
+from django.contrib.auth.models import User
+from django.core.exceptions import ObjectDoesNotExist
+from django.db.models import Q
+from taskcluster.sync import Auth
+from taskcluster.utils import scope_match
+
+logger = logging.getLogger(__name__)
+
+
+class TaskclusterAuthBackend(object):
+    """
+        result of tc_auth.authenticateHawk has the form:
+
+        {'status': 'auth-success',
+         'scopes': ['assume:mozilla-group:ateam',
+                    'assume:mozilla-group:vpn_treeherder',
+                    'assume:mozilla-user:biped@mozilla.com',
+                    'assume:mozillians-user:biped',
+                    ...
+                    'assume:project-admin:ateam',
+                    'assume:project-admin:treeherder',
+                    'assume:project:ateam:*',
+                    'assume:project:treeherder:*',
+                    'assume:worker-id:*',
+                    'secrets:set:project/treeherder/*'],
+         'scheme': 'hawk',
+         'clientId': 'mozilla-ldap/biped@mozilla.com',
+         'expires': '2016-10-31T17:40:45.692Z'}
+    """
+
+    def _get_scope_value(self, result, scope_prefix):
+        for scope in result["scopes"]:
+            if scope.startswith(scope_prefix):
+                return scope[len(scope_prefix):]
+        return None
+
+    def _get_email(self, result):
+        """
+        Get the user's email from the mozilla-user scope
+
+        """
+        email = self._get_scope_value(result, "assume:mozilla-user:")
+
+        if not email or email == "*":
+            # try finding the email in the clientId.  It can then be used to
```

Comment:
This doesn't seem secure - we should fail authentication in `not email or email == "*"` case surely? I think it's reasonable to just say we don't support !LDAP people for now. (The description text on https://login.taskcluster.net/ says that non-employees can still use the Okta login option).

@djmitche, what are possible values for the `assume:mozilla-user:...` scope? Can a TC admin create arbitrary ones of these? (Is that what the `"*"` value is about?) Having to support (ie skip) missing email addresses is fine, but we absolutely do need to have a secure way to get a users email address, if it's present for a set of credentials.

Even with the previous version of the PR (that directly used the sheriff scope to set permissions), this would have been an issue, since we need to use the email address to migrate existing users. Admin/sheriff permission bits aren't the only sensitive items that need migrating, we also have Hawk API secrets in the Treeherder DB that are owned by individual users (that can't be represented by TC scopes). If someone can spoof the `assume:mozilla-user:...` scope, then they can view the Hawk secret for any API credentials owned by that user.